### PR TITLE
Add port for gnu scientific library

### PR DESCRIPTION
**Pull request by muemart**

### CODE REVIEW

File: ports/libgsl/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,71 @@
+cmake_minimum_required(VERSION 3.8)
+project(libgsl C)
+
+option(INSTALL_HEADERS "Install public header files" ON)
+set(WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
Comment (traversaro) [inline ports/libgsl/CMakeLists.txt]: This seems to be unused and can be removed, right?

File: ports/libgsl/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+if(INSTALL_HEADERS)
+    set_target_properties(libgsl PROPERTIES PUBLIC_HEADER "${PUBLIC_HEADERS}")
+endif()
+target_include_directories(libgsl PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
+# For the build, we need to copy all headers to the gsl directory
+file(COPY ${PUBLIC_HEADERS} DESTINATION "${CMAKE_CURRENT_SOURCE_DIR}/gsl")
+
+if(BUILD_SHARED_LIBS)
+    target_compile_definitions(libgsl PRIVATE -DDLL_EXPORT)
```
Comment (traversaro) [inline ports/libgsl/CMakeLists.txt]: You can use directly the [`DEFINE_SYMBOL`](https://cmake.org/cmake/help/v3.7/prop_tgt/DEFINE_SYMBOL.html) CMake target property to define `DLL_EXPORT` only when building gsl as a shared library, i.e. : 
~~~
set_target_properties(libgsl PROPERTIES DEFINE_SYMBOL DLL_EXPORT)
~~~

File: ports/libgsl/CMakeLists.txt (hunk 1 above)
Comment (muemart) [inline ports/libgsl/CMakeLists.txt]: No, this is supposed to export all symbols of the dll automatically. The code only has __declspecs for variables. But I just noticed I did this the wrong way.

File: ports/libgsl/0002-add-fp-control.patch [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
++#include <config.h>
++#include <gsl/gsl_ieee_utils.h>
++#include <gsl/gsl_errno.h>
++
++const char *fp_env_string = "round-to-nearest,double-precision,mask-all";
++
++int
++gsl_ieee_set_mode (int precision, int rounding, int exception_mask)
++{
++	unsigned int old, mode = _DN_SAVE, mask = _MCW_DN | _MCW_RC | _MCW_EM;
```
Comment (KindDragon) [inline ports/libgsl/0002-add-fp-control.patch]: Can you use 2 space indentation like in other libgsl code?

**Comment by ras0219-msft** [comment]
Could you rename this to `gnusl` or `libgnusl` instead of `libgsl`?

We would like to avoid overloading any library names on the prefix `lib` since it's pretty ambiguous for most projects whether they will have a `lib` suffix. In the future, we may look into aliasing `libX` to `X` as well (see #540).

**Comment by muemart** [comment]
Do you mean the port, the library, or both? I think libgsl is the normal name of the library in UNIX-land, so renaming the library might be surprising for some, but not renaming still has the same problem.

**Comment by ras0219-msft** [comment]
Just the port should be fine. We may have trouble if the Guidelines Support Library ever becomes non-header-only, but we'll deal with that later (they'll also have to come up with a solution for UNIX in that case, so they'll probably not name the library `libgsl` 😄).

**Comment by muemart** [comment]
Okay, I renamed to port to gnusl.

**Comment by ras0219-msft** [comment]
Thanks!

Is it possible to get the fp-control patch upstreamed?

**Comment by muemart** [comment]
That patch has been floating around the internet for ages apparently, and every single gsl port for Windows uses this. I don't know why it hasn't been upstreamed, but it isn't my code.

**Comment by ras0219-msft** [comment]
As a notice:

I discussed the naming collision problem with Neil and we determined the best course of action is to rename this port back to just `gsl`, while renaming the Guidelines Support Library to `ms-gsl`.

Please do know that your quick rename was still very appreciated :).

**Pull request merged**